File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/objects/ColourableObject.java
package pw.switchcraft.plethora.gameplay.modules.glasses.objects;

import net.minecraft.network.PacketByteBuf;

import javax.annotation.Nonnull;

public abstract class ColourableObject extends BaseObject implements Colourable {
	private int colour = DEFAULT_COLOUR;

	public ColourableObject(int id, int parent, byte type) {
		super(id, parent, type);
	}

	@Override
	public final int getColour() {
		return colour;
	}

	@Override
	public final void setColour(int colour) {
		if (this.colour != colour) {
			this.colour = colour;
			setDirty();
		}
	}

	protected int getRed() {
		return (colour >> 24) & 0xFF;
	}

	protected int getGreen() {
		return (colour >> 16) & 0xFF;
	}

	protected int getBlue() {
		return (colour >> 8) & 0xFF;
	}

	protected int getAlpha() {
		return colour & 0xFF;
	}

	@Override
	public void writeInitial(@Nonnull PacketByteBuf buf) {
		buf.writeInt(colour);
	}

	@Override
	public void readInitial(@Nonnull PacketByteBuf buf) {
		colour = buf.readInt();
	}
}

File: src/main/java/io/sc3/plethora/gameplay/modules/glasses/objects/ColourableObject.kt
package pw.switchcraft.plethora.gameplay.modules.glasses.objects

import net.minecraft.network.PacketByteBuf
import pw.switchcraft.plethora.util.DirtyingProperty

abstract class ColourableObject(id: Int, parent: Int, type: Byte) : BaseObject(id, parent, type), Colourable {
  override var colour by DirtyingProperty(DEFAULT_COLOUR.toInt())

  val red: Int
    get() = colour shr 24 and 0xFF
  val green: Int
    get() = colour shr 16 and 0xFF
  val blue: Int
    get() = colour shr 8 and 0xFF
  val alpha: Int
    get() = colour and 0xFF

  override fun writeInitial(buf: PacketByteBuf) {
    buf.writeInt(colour)
  }

  override fun readInitial(buf: PacketByteBuf) {
    colour = buf.readInt()
  }
}
